### PR TITLE
[VL] Add a config for shuffle writer buffer size to avoid OOM

### PR DESCRIPTION
**Pull request by zuochunwei**
## What changes were proposed in this pull request?

make the default config to 4096, we can adjust to a small value in case of a low memory test system. 

(Please fill in changes proposed in this fix)

(Fixes: \#ISSUE-ID)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 1]
```diff
@@ -785,6 +791,12 @@ object GlutenConfig {
       .intConf
       .createWithDefault(4096)
 
+  val SHUFFLE_WRITER_BUFFER_SIZE =
+    buildConf(GLUTEN_SHUFFLE_WRITER_BUFFER_SIZE)
+      .internal()
+      .intConf
+      .createWithDefault(4096)
+
```
Comment (zhztheplayer) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:799]: Can we by default use `spark.gluten.sql.columnar.maxBatchSize`? For backward compatibility.

File: docs/Configuration.md [hunk 2]
```diff
@@ -44,6 +44,7 @@ You can add these configurations into spark-defaults.conf to enable or disable t
...
 | spark.gluten.sql.columnar.maxBatchSize                  | Set the number of rows for the output batch                                                                                                                                                                                                                                                                                                                                                                                                          | 4096                                                 |
+| spark.gluten.shuffleWriterBufferSize                  | Set the number of buffer rows for the shuffle writer                                                                                                                                                                                                                                                                                                                                                                                                          | 4096                                                 |
```
Comment (winningsix) [inline docs/Configuration.md:47]: nit: ```spark.gluten.shuffleWriter.BufferSize``` better? Similar name convention to ```spark.shuffle.file.buffer``` in Spark.

Comment (zuochunwei) [reply docs/Configuration.md:47]: OK

**Comment by zhztheplayer** [comment]
I've update the PR for default value of the option. @zuochunwei

**Comment by zuochunwei** [comment]
> I've updated the PR for default value of the option. @zuochunwei

thanks for your revision.

**Review by zhztheplayer — APPROVED** [review]

**Comment by zhztheplayer** [comment]
@marin-ma Would you like to take a look? Thanks.

**Comment by marin-ma** [comment]
LGTM. Thanks!

**Review by marin-ma — APPROVED** [review]

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 3]
```diff
@@ -333,6 +337,9 @@ object GlutenConfig {
   // Batch size.
   val GLUTEN_MAX_BATCH_SIZE_KEY = "spark.gluten.sql.columnar.maxBatchSize"
 
+  // Shuffle Writer buffer size.
+  val GLUTEN_SHUFFLE_WRITER_BUFFER_SIZE = "spark.gluten.shufflewriter.bufferSize"
```
Comment (zjuwangg) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:341]: "spark.gluten.shufflewriter.bufferSize" --> better to rename to  "spark.gluten.shuffleWriter.bufferSize"

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxTPCDSSuite.scala [hunk 4]
```diff
@@ -57,6 +57,7 @@ class VeloxTPCDSSuite extends WholeStageTransformerSuite {
       .set("spark.sql.sources.useV1SourceList", "avro")
       .set("park.sql.adaptive.enabled", "true")
       .set("spark.gluten.sql.columnar.maxBatchSize", "4096")
+      .set("spark.gluten.shuffleWriter.bufferSize", "4096")
```
Comment (zjuwangg) [inline backends-velox/src/test/scala/io/glutenproject/execution/VeloxTPCDSSuite.scala:60]: The key is inconsistent with   GLUTEN_SHUFFLE_WRITER_BUFFER_SIZE for the case-sensitive problem

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala (hunk 3 above)
Comment (zjuwangg) [reply shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:341]: I opened PR to fix this typo https://github.com/oap-project/gluten/pull/3073

**Pull request merged**